File: ios/VEP/Models/ContactLog.swift
import Foundation

/// Contact log model representing a voter contact event
struct ContactLog: Codable, Identifiable {
    let id: UUID
    let assignmentId: UUID
    let voterId: UUID
    let contactType: ContactType
    let result: String?
    let supportLevel: Int?
    let location: Coordinate
    let contactedAt: Date
    
    enum CodingKeys: String, CodingKey {
        case id
        case assignmentId = "assignment_id"
        case voterId = "voter_id"
        case contactType = "contact_type"
        case result
        case supportLevel = "support_level"
        case location
        case contactedAt = "contacted_at"
    }
    
    /// Initialize a new contact log (for creating new contacts)
    init(id: UUID = UUID(),
         assignmentId: UUID,
         voterId: UUID,
         contactType: ContactType,
         result: String?,
         supportLevel: Int?,
         location: Coordinate,
         contactedAt: Date = Date()) {
        self.id = id
        self.assignmentId = assignmentId
        self.voterId = voterId
        self.contactType = contactType
        self.result = result
        self.supportLevel = supportLevel
        self.location = location
        self.contactedAt = contactedAt
    }
}

/// Contact type enumeration
enum ContactType: String, Codable, CaseIterable {
    case knocked
    case phone
    case text
    case email
    case notHome = "not_home"
    case refused
    case moved
    case deceased
    
    var displayName: String {
        switch self {
        case .knocked: return "Knocked"
        case .phone: return "Phone"
        case .text: return "Text"
        case .email: return "Email"
        case .notHome: return "Not Home"
        case .refused: return "Refused"
        case .moved: return "Moved"
        case .deceased: return "Deceased"
        }
    }
    
    var isSuccessfulContact: Bool {
        switch self {
        case .knocked, .phone, .text, .email:
            return true
        case .notHome, .refused, .moved, .deceased:
            return false
        }
    }
}

File: backend/app/models/contact_log.py
"""
VEP MVP Backend - Contact Log Models

Database models and Pydantic schemas for contact logs.
"""

from datetime import datetime
from typing import Optional
from uuid import UUID, uuid4

from sqlmodel import Field, SQLModel

from app.models.voter import Coordinate


class ContactType:
    """Contact type constants."""
    KNOCKED = "knocked"
    PHONE = "phone"
    TEXT = "text"
    EMAIL = "email"
    NOT_HOME = "not_home"
    REFUSED = "refused"
    MOVED = "moved"
    DECEASED = "deceased"


class ContactLogBase(SQLModel):
    """Base contact log fields shared across schemas."""
    assignment_id: UUID
    voter_id: UUID
    contact_type: str
    result: Optional[str] = None
    support_level: Optional[int] = Field(default=None, ge=1, le=5)


class ContactLog(ContactLogBase, table=True):
    """
    ContactLog database model.
    
    Records every voter contact interaction made by canvassers.
    Tracks contact type, result, support level changes, and GPS location.
    Location field records where the canvasser was when logging (for verification).
    """
    __tablename__ = "contact_logs"
    
    id: UUID = Field(default_factory=uuid4, primary_key=True)
    user_id: UUID = Field(foreign_key="users.id")
    # Note: location is stored as GEOMETRY(POINT, 4326) in PostgreSQL
    # We'll handle conversion in the routes layer
    contacted_at: datetime = Field(default_factory=datetime.utcnow)
    created_at: datetime = Field(default_factory=datetime.utcnow)


class ContactLogCreate(ContactLogBase):
    """Schema for creating a new contact log."""
    location: Optional[Coordinate] = None


class ContactLogUpdate(SQLModel):
    """Schema for updating a contact log."""
    contact_type: Optional[str] = None
    result: Optional[str] = None
    support_level: Optional[int] = Field(default=None, ge=1, le=5)


class ContactLogRead(ContactLogBase):
    """Schema for reading a contact log."""
    id: UUID
    user_id: UUID
    location: Optional[Coordinate] = None
    contacted_at: datetime
    created_at: datetime


class ContactLogWithDetails(ContactLogRead):
    """Schema for reading a contact log with voter details."""
    voter_name: Optional[str] = None
